src/rsa/blinding.rs

```diff
@@ -0,0 +1,585 @@
+//
+// However, for simplicity, we only support this for the case when the number
+// of bits in the bound (/ public key modulus) is a multiple of LIMB_BITS,
+// or one less, which we expect to be the case in performance-sensitive
+// applications, where, e.g., the 2048 or 2047-bit modulus will be the product
+// of two 1024-bit integers.
+fn select_sampling_params(max_exclusive: &[Limb]) -> SamplingParams {
+    let most_sig = max_exclusive[max_exclusive.len() - 1];
+
+    if most_sig >> (LIMB_BITS - 3) == 0b100 {
+        SamplingParams {
+            most_sig_limb_mask: 1, // effectively a carry over into a new, more-significant limb
+            reduce_when_over_bound: true,
+            extend_limbs_by_one: true,
+        }
+    } else if most_sig >> (LIMB_BITS - 4) == 0b0100 {
+        SamplingParams {
+            most_sig_limb_mask: Limb::max_value(),
+            reduce_when_over_bound: true,
+            extend_limbs_by_one: false,
+        }
+    } else {
+        SamplingParams {
+            most_sig_limb_mask: Limb::max_value() >> most_sig.leading_zeros(),
+            reduce_when_over_bound: false,
+            extend_limbs_by_one: false,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {core,rand};
+    use limb::*;
+
+    #[test]
+    fn test_select_sampling_params() {
+        use super::select_sampling_params;
+
+        let starting_with_0b100 = &[
+            1 << (LIMB_BITS - 1),
+            1 << (LIMB_BITS - 1) | 1,
+            (1 << (LIMB_BITS - 1)) | (1 << LIMB_BITS - 4),
+            (1 << (LIMB_BITS - 1)) | (Limb::max_value() >> 3),
+        ];
+
+        for l in starting_with_0b100 {
+            let x = [*l];
+            let p = select_sampling_params(&x[..]);
+            assert!(p.extend_limbs_by_one);
+            assert!(p.reduce_when_over_bound);
+            assert_eq!(1, p.most_sig_limb_mask);
+        }
+
+        let starting_with_0b0100 = &[
+            1 << (LIMB_BITS - 2),
+            1 << (LIMB_BITS - 2) | 1,
+            (1 << (LIMB_BITS - 2)) | (1 << LIMB_BITS - 5),
+            (1 << (LIMB_BITS - 2)) | (Limb::max_value() >> 4),
+        ];
+
+        for l in starting_with_0b0100 {
+            let x = [*l];
+            let p = select_sampling_params(&x[..]);
+            assert!(!p.extend_limbs_by_one);
+            assert!(p.reduce_when_over_bound);
+            assert_eq!(Limb::max_value(), p.most_sig_limb_mask);
+        }
+
+        macro_rules! assert_normal {
+            ($i:expr, $l:expr) => {
+                {
+                    let x = [$l];
+                    let p = select_sampling_params(&x[..]);
+                    let mask = Limb::max_value() >> (LIMB_BITS - 1 - $i);
+                    assert!(!p.extend_limbs_by_one);
+                    assert!(!p.reduce_when_over_bound);
+                    assert_eq!(mask, p.most_sig_limb_mask);
+                }
+            }
+        }
+
+        for i in 0..(LIMB_BITS - 2) {
```

Comment:
Add a comment explaining why you stoped at `LIMB_BITS - 2`. The reason is that you want to ensure that the value never starts with `0b100` or `0b0100`.